### PR TITLE
add unit tests for OAuth2Utils

### PR DESCRIPTION
**Pull request by ChanikaRuchini**
Added more unit tests for OAuth2Utils.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java [hunk 1]
```diff
@@ -415,9 +432,19 @@ public void testIsHashDisabled() {
         assertEquals(OAuth2Util.isHashDisabled(), false);
     }
 
+    @Test
+    public void testIsHashEnabled() {
+
+        when(OAuthServerConfiguration.getInstance().isClientSecretHashEnabled()).thenReturn(true);
+
+        assertEquals(OAuth2Util.isHashEnabled(), true);
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:440]: use `assertTrue` here

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java [hunk 2]
```diff
@@ -1799,4 +1966,300 @@ public void testValidateRequestTenantDomain(boolean isTenantQualifiedURLsEnabled
 
         assertEquals(isInvalidClientExceptionThrown, isExceptionExpected);
     }
+
+    @DataProvider(name = "createResponseType")
+    public Object[][] createResponseType() {
+
+        return new Object[][]{
+                {"", false},
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:1974]: can also pass another data set where responseType is null

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java [hunk 3]
```diff
@@ -1799,4 +1966,300 @@ public void testValidateRequestTenantDomain(boolean isTenantQualifiedURLsEnabled
...
+                {OAuthConstants.CODE_IDTOKEN, true},
+                {OAuthConstants.CODE_IDTOKEN_TOKEN, true}
+        };
+    }
+
+    @Test(dataProvider = "createResponseType")
+    public void testisHybridResponseType(String responseType, boolean expected) {
+
+        assertEquals(OAuth2Util.isHybridResponseType(responseType), expected);
+
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:1985]: Unnecessary new line

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java [hunk 4]
```diff
@@ -1799,4 +1966,300 @@ public void testValidateRequestTenantDomain(boolean isTenantQualifiedURLsEnabled
...
+        List<String> result = OAuth2Util.getOIDCScopes(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME);
+        assertEquals(result.size(), 0);
+    }
+
+    @DataProvider(name = "tokenTypeProvider")
+    public static Object[][] tokenTypeProvider() {
+
+        return new Object[][]{
+                {"dummyTokenType"},
+                {null}};
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:2024]: consider empty string as well

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java [hunk 5]
```diff
@@ -1799,4 +1966,300 @@ public void testValidateRequestTenantDomain(boolean isTenantQualifiedURLsEnabled
...
+    }
+
+    @DataProvider(name = "authenticatedIDPProvider")
+    public static Object[][] authenticatedIDPProvider() {
+
+        String authenticatedIDP = "dummyAuthenticatedIDP";
+        return new Object[][]{
+                {authenticatedIDP, OAuthConstants.UserType.FEDERATED_USER_DOMAIN_PREFIX +
+                        OAuthConstants.UserType.FEDERATED_USER_DOMAIN_SEPARATOR + authenticatedIDP},
+                {null, OAuthConstants.UserType.FEDERATED_USER_DOMAIN_PREFIX}};
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:2105]: consider empty string as well

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java [hunk 6]
```diff
@@ -1799,4 +1966,300 @@ public void testValidateRequestTenantDomain(boolean isTenantQualifiedURLsEnabled
...
+    @Test
+    public void testGetTokenBindingReference() {
+
+        assertNotNull(OAuth2Util.getTokenBindingReference("dummyReference"));
+    }
+
+    @Test
+    public void testGetTokenBindingReferenceWithNullValue() {
+
+        assertNull(OAuth2Util.getTokenBindingReference(null));
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:2123]: Better to write a test case for `OAuth2Util.getTokenBindingReference("")` as well

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java [hunk 7]
```diff
@@ -415,9 +433,19 @@ public void testIsHashDisabled() {
         assertEquals(OAuth2Util.isHashDisabled(), false);
     }
 
+    @Test
+    public void testIsHashEnabled() {
+
+        when(OAuthServerConfiguration.getInstance().isClientSecretHashEnabled()).thenReturn(true);
+
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:440]:
```suggestion
```

**Review by dewniMW — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java [hunk 8]
```diff
@@ -1799,4 +1969,315 @@ public void testValidateRequestTenantDomain(boolean isTenantQualifiedURLsEnabled
...
+                {"", false},
+                {null, false},
+                {OAuthConstants.CODE_TOKEN, true},
+                {OAuthConstants.CODE_IDTOKEN, true},
+                {OAuthConstants.CODE_IDTOKEN_TOKEN, true}
+        };
+    }
+
+    @Test(dataProvider = "createResponseType")
+    public void testisHybridResponseType(String responseType, boolean expected) {
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:1986]: Fix improper camel casing in method name  `testisHybr...` -> `testIsHybr...`.

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java [hunk 9]
```diff
@@ -1799,4 +1969,315 @@ public void testValidateRequestTenantDomain(boolean isTenantQualifiedURLsEnabled
...
+        accessTokenDO.setConsumerKey("dummyConsumerKey");
+
+        mockStatic(OAuthCache.class);
+        when(OAuthCache.getInstance()).thenReturn(oAuthCacheMock);
+        if (isCacheAvailable) {
+            when(oAuthCacheMock.getValueFromCache(any(OAuthCacheKey.class))).thenReturn(accessTokenDO);
+        } else {
+            WhiteboxImpl.setInternalState(OAuthTokenPersistenceFactory.getInstance(), "tokenDAO", accessTokenDAO);
+            when(accessTokenDAO.getAccessToken(anyString(), anyBoolean())).thenReturn(accessTokenDO);
+
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:2187]: Remove unnecessary new line.

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java (hunk 7 above)
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:440]: Remove redundant line

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java [hunk 10]
```diff
@@ -1496,8 +1599,76 @@ public void testIsPKCESupportEnabled() throws Exception {
...
+    public Object[][] createPKCEData() {
+
+        String verificationCode = generateCodeVerifier(77);
+        return new Object[][]{
+                {verificationCode, verificationCode, OAuthConstants.OAUTH_PKCE_PLAIN_CHALLENGE, true},
+                {"dummyReferenceCodeChallenge", verificationCode, OAuthConstants.OAUTH_PKCE_PLAIN_CHALLENGE, false},
+                {"dummyReferenceCodeChallenge", verificationCode, OAuthConstants.OAUTH_PKCE_S256_CHALLENGE, false},
+                {verificationCode, verificationCode, null, true},
+                {"", null, OAuthConstants.OAUTH_PKCE_S256_CHALLENGE, true},
+
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:1612]: Remove redundant line

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java (hunk 8 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:1986]: Addressed with commit [9e575a3](https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1691/commits/9e575a323c0604e9985d430a5b3fee4942d86cb1)

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java (hunk 9 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:2187]: Addressed with commit [9e575a3](https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1691/commits/9e575a323c0604e9985d430a5b3fee4942d86cb1)

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java (hunk 7 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:440]: Addressed with commit [9e575a3](https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1691/commits/9e575a323c0604e9985d430a5b3fee4942d86cb1)

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java (hunk 10 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/OAuth2UtilTest.java:1612]: Addressed with commit [9e575a3](https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1691/commits/9e575a323c0604e9985d430a5b3fee4942d86cb1)

**Pull request merged**